Improve Modular Code pattern

fioddor opened the pull request:
- Problem Generalize the problem (less contextual). Modularization solves more problems in different contexts.
- Context
  - Remove specific circumstances
  - Add generic ones
- Forces:
  - Group in against / in favour / neutral
  - Turn 'Frequent turnover' into a favorable circumstance
  - Explain impact opportunity for newbies
- Solutions:
  - Group and order advice bits
  - Add: education on why
  - Add advice on finding modular feature candidates
  - Add advice to address bug changes in small bits
- Resulting Context
  - Move bits from solutions section here (they describe the effects, not the undertakings)

Commit:
Focus the problem

Generalize the problem (less contextual):
- Predisposition is more of a force than a problem.
- Modularization solves more problems in different contexts.

Detail further the effects of lack of modularization.

Commit:
Correct and expand the context

- Remove the lack of mandate because it might or not be the case.
- Remove the exclusion of legacy code. It is a clear use case.
- Making code modular doesn't take extra effort and time over the
  alternatives.
+ Include advantages for documentation

Commit:
Organize forces

* Group in pro/con/neutral
* Turn Frequent turnover into a pro
* Explain impact opportunity for newbies

Commit:
Describe resulting context

Move some points from Solutions section to Resulting Context.
They describe effects of modularization, not actions to be taken
towards it.

Commit:
Reorganize solutions

1. Describe modularization as the main solution. The rest are secondary
   actions to implement and/or support modularization as a practice.
2. Prepare (offer education) first
3. Then, decide to go for it
4. Finally take the first steps.

- Mitigating risk and fear of changes is a result, not an action.
- Providing incentives is questionable. Its benefits should suffice.
- Paralell labour moved to 'Resulting context' section already as
  it is a result, not an action. Remove duplication.

Commit:
Extend solutions

Extend the first steps.
* Group standards for reusability
* Advice small steps

Commit:
Extend solutions

+ Extend the education part.
+ Structure it

Commit:
Fix lint advice

NewMexicoKid commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,66 @@ The lack of modularization in the software architecture prevents reuseability, a
 
-* There is no corporate mandate for modularized code.
-* This is a new product/new code, not a legacy product/code.
+* There might be potentially reusable functionality buried in monolythic code.
```
monolithic

spier replied:
Hi @NewMexicoKid. Not sure if you have seen that GitHub feature yet.

When commenting you can hit the ![Screenshot 2022-10-26 at 18 09 58 ](https://user-images.githubusercontent.com/163029/198078602-5dd1f1b0-e9cc-4c6f-b80b-e25748e7def8.png)  symbol at the top of the comment box and suggest specific changes.
That allows the authors to clearly understand what you are referring to. Plus if the change doesn't need further conversation they can just click "commit" and are done :)

Below a screenshot of the button that I am referring to:

![Screenshot 2022-10-26 at 18 08 05 ](https://user-images.githubusercontent.com/163029/198078268-07378837-6f99-4cb5-9dab-c200d0b587bb.png)

mishari commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,66 @@ The lack of modularization in the software architecture prevents reuseability, a
+  1. define decoupling scope and prefferred decoupling methods,
+  1. develop automated testing protection
+  1. document usage of modules.
 
```
I spoke to @fioddor about this and I wish to add one more line here which I think is super useful
```suggestion
* More information and techniques are available from the book Working Effectively with Legacy Code by Michael Feathers and Refactoring: Improving the Design of Existing Code by Martin Fowler

```

spier committed:
Add book references

Co-authored-by: Mishari Muqbil <mishari@mishari.net>

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,67 @@ The lack of modularization in the software architecture prevents reuseability, a
+  1. define decoupling scope and prefferred decoupling methods,
+  1. develop automated testing protection
+  1. document usage of modules.
+* More information and techniques are available from the book Working Effectively with Legacy Code by Michael Feathers and Refactoring: Improving the Design of Existing Code by Martin Fowler
```
```suggestion
* More information and techniques are available from the book **Working Effectively with Legacy Code** by Michael Feathers and **Refactoring: Improving the Design of Existing Code** by Martin Fowler
```

Add some formatting for the book references

spier committed:
Add some formatting for the book references

mishari commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,67 @@ The lack of modularization in the software architecture prevents reuseability, a
+  1. find internal would-be customers. Better ask them over guessing on your own.
+  1. define decoupling scope and prefferred decoupling methods,
+  1. develop automated testing protection
+  1. document usage of modules.
```
```suggestion
  1. document usage of modules.
  1. introduce semantic versioning
  1. use a dependency management software such as pip, nuget, npm
```

mishari reviewed (changes requested):
Many orgs don't realize semantic versioning and dependency management also applies to their internal code base. I propose these lines so people can look up more information.

Commit:
Add more steps to modularization

Add more steps to modularization sequence.

Co-authored-by: Mishari Muqbil <mishari@mishari.net>

Commit:
Fix typo

Thanks @NewMexicoKid

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+* With rare exceptions, monolithinc code is difficult to contribute to.
+* The same functionality is reimplemented redundantly causing
+  * Waste of unneeded re-design and re-development
+  * Duplicated mainteinance effort
```
```suggestion
With rare exceptions, monolithic code is difficult to contribute to and extend.
That leads development teams to build similar functionality multiple times, leading to both waste (unneeded re-design and re-development) as well as duplicated maintenance effort.
```

spier replied:
In terms of formatting, I am suggesting to not use bullets here. We already have enough bullets in the other sections :)

Also fixed some typos and rewrote the Problem statement a bit.

fioddor replied:
Typos and formatting: ok

I see that _monolithic code_ might be a cause for _redundant implementation_, but it is not always so.  Redundant implementation might happen for other reasons, so I'd keep it as another problem that modularization solves.

spier replied:
Sorry but I could not follow your train of thought here, please help me :)

Are you saying that your lines 11 and 12 above are describing two distinct unrelated problems, that can be improved through more modular code?

fioddor replied:
Yes, 2 different problems:

1. Monolithic code packing much functionality makes it difficult to contribute (much to read, much to grasp, much to test, much to review, and much to document). Finding slots and the right mood to attack big chunks of work is difficult. This is regardless of how many other alternative implementations you have.
2. Redundant implementations tend to be a waste of effort developing, documenting, and maintaining them all.

Strictly speaking, modularity is about reusing parts (modules) and solves redundancy.

But engineering the parts to be small and narrow-focused (breaking the monolith) helps a lot in fostering collaboration. More so in the case of volunteer collaboration.

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
 ## Resulting Context
 
-**Pattern Idea**: Time is spent making the shared code modular so it can be reused.  
+* Modular code facilitates division of paralell labour (as required for work in larger teams, or as an open source project). Modular projects enable team members to focus on their own smaller assigned tasks.
```
```suggestion
* Modular code facilitates division of parallel labor (as required for work in larger teams, or as an open source project).
* Modular projects enable team members to focus on their own smaller assigned tasks.
```

Fixed spelling in first sentence. Also changed "labor" to American English.

Would it make sense to break out the 2nd sentence into its own bullet?

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
 
-**Pattern Idea**: Time is spent making the shared code modular so it can be reused.  
+* Modular code facilitates division of paralell labour (as required for work in larger teams, or as an open source project). Modular projects enable team members to focus on their own smaller assigned tasks.
+* Modular code makes Agile implementation easier.
```
What do you mean by "Agile implementation" in this context? e.g. it could mean something similar to the "smaller assigned tasks" above but I am not sure.

fioddor replied:
Smaller grains make sand more fluid.

Agile relies on small iterations. Smaller codebase bits and atomic unit tests provide more alternative combinations which in turn allows to pack more work into timeboxed sprints. It also helps when replanning. And smaller bits have smaller scopes (less to learn and document), so user stories get more specific (less to discuss).

For the same functionality, you'll need to change more bits, but the modularity helps to materialize sooner (earlier and more reliable validations) and to prioritize the most relevant bits (separate value from waste).

fioddor replied:
It's all about smaller tasks, but the value provided by making tasks smaller is different.

spier replied:
> Smaller grains make sand more fluid.

⬆️ love that! Had not heard that expression before but it is a great mental picture.

I can see that it might be hard to pack your whole statement into row 72. Still wondering if we could add some of it? e.g.

> Modular code makes Agile implementation easier e.g. by allowing for smaller iterations and through that earlier customer feedback.

spier replied:
Saw the changes the Agile bit. Fantastic!

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+  * You might be dealing with legacy systems. Might be a fear that if not done properly, quality might be impacted.
+  * Architectural constraints might impact modularity.
+  * The value of modularity is nearly invisible for end users.
+* In favour:
```
```suggestion
* In favor:
```

Switching to American English (AE)

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+  Legacy code tends to be less modular than new one.
 * There is an available repository for sharing code.
-* Making code modular takes extra effort and time to develop.
+* Source code can be a useful as documentation on small enough pieces.
```
```suggestion
* Source code can be as useful as documentation on small enough pieces.
```

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+  * On why: Lack of modularity is technical debt. It increases costs but hides them by shifting these into the future.
+  * On how: Modular code is well-understood; there is a lot of literature in favor of this.
+  * Select certain "success projects", teams that will develop reusable code and demonstrate the long term success. This can help motivate others (they see what is possible and what is in it for them). Transparency is critical.
+* Introduce refactoring as organic effort and strive for it to become part of the culture.
```
```suggestion
* Introduce refactoring as an organic effort and strive for it to become part of the culture.
```

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+  * On how: Modular code is well-understood; there is a lot of literature in favor of this.
+  * Select certain "success projects", teams that will develop reusable code and demonstrate the long term success. This can help motivate others (they see what is possible and what is in it for them). Transparency is critical.
+* Introduce refactoring as organic effort and strive for it to become part of the culture.
+  * Acknowledge and drive acceptance for the investion of modularization. Build time into the release schedule for this.
```
```suggestion
  * Acknowledge and drive acceptance for the investment in modularization. Build time into the release schedule for this.
```

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,69 @@ The lack of modularization in the software architecture prevents reuseability, a
+    Establish standards on testing methodology, labeling of repos.
+* Eat the elephant in small bites. Break modularization down into baby steps, like
+  1. find internal would-be customers. Better ask them over guessing on your own.
+  1. define decoupling scope and prefferred decoupling methods,
```
```suggestion
  1. define decoupling scope and preferred decoupling methods,
```

spier reviewed (commented):
@fioddor I did a first round of review. Great job!

Left a bunch of comments, mostly related to some spelling errors.

I had a hard time assessing how the old and new Solution section are different, as so much was changed there. Will read it one more time in the coming days, problem just focusing on whether the new Solution section makes sense for me (rather than comparing it to the old one).

Btw by all means add yourself to the "Authors" section, as you are doing a rather significant rewrite here.

Commit:
fix typo

Co-authored-by: Sebastian Spier <github@spier.hu>

Commit:
Correct weird vocabulary

Co-authored-by: Sebastian Spier <github@spier.hu>

Commit:
Fix typo and spin off a point

Co-authored-by: Sebastian Spier <github@spier.hu>

Commit:
Minor fixes

Co-authored-by: Sebastian Spier <github@spier.hu>

Commit:
Reword the problem

According to review comments. Thank you, Sebastian Spier

Commit:
Elaborate on agile

Elaborate on how modular code helps agile.

Commit:
Add myself as author

As suggested by my reviewer.

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
+* Offer education.
+  * On why: Lack of modularity is technical debt. It increases costs but hides them by shifting these into the future.
+  * On how: Modular code is well-understood; there is a lot of literature in favor of this.
+  * Select certain "success projects", teams that will develop reusable code and demonstrate the long term success. This can help motivate others (they see what is possible and what is in it for them). Transparency is critical.
```
```suggestion
  * Select certain "success projects", teams that have developed reusable code and can demonstrate the long term success. This can help motivate others, as they see what is possible and what is in it for them. Learning directly from other developers is critical.
```

Changes:

* Would you select teams here that have already applied modularization successfully in the past, or rather some that want to do so on the future? I suspect the former, and have changed the tense accordingly.
* Removed the brackets
* Changed the last sentence significantly here. Wondering if this is more precise than "transparency"?
 
This is all your text though, so you decide if you like any of this.

spier replied:
✅

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
+
+* Offer education.
+  * On why: Lack of modularity is technical debt. It increases costs but hides them by shifting these into the future.
+  * On how: Modular code is well-understood; there is a lot of literature in favor of this.
```
```suggestion
  * On how: Modular code is well-understood; there is a lot of literature in favor of this that can be used for training.
```

spier replied:
✅

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
-* There is no corporate mandate for modularized code.
-* This is a new product/new code, not a legacy product/code.
+* There might be potentially reusable functionality buried in monolithic code.
+  Legacy code tends to be less modular than new one.
```
Is this meant to be it's own bullet, or should it be part of the previous one?

If the latter, then we need to check after publishing this whether gitbook renders this correctly. I vaguely recall some behavior there where GitHub and gitbook rendering of markdown differ.

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
+
+The same functionality is reimplemented redundantly causing waste of unneeded re-design and re-development and duplicated mainteinance effort.
+
+Sometimes the first problem causes the second, but it might also happen spontaneously.
```
👍

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
+  1. develop automated testing protection
+  1. document usage of modules.
+  1. introduce semantic versioning
+  1. use a dependency management software such as pip, nuget, npm
```
```suggestion
  1. Find internal would-be customers. Better ask them over guessing on your own.
  1. Define decoupling scope and preferred decoupling methods.
  1. Develop automated testing protection.
  1. Document usage of modules.
  1. Introduce semantic versioning.
  1. Use a dependency management software such as pip, nuget, npm.
```

I think we should start all points with capital letters, to be consistent with the rest of the pattern.
Also added punctuation after each bullet point.

One question:

This is the only part of the pattern where we use a numbered list.
Is this on purpose? i.e. is there an order to this?

spier replied:
@fioddor This is the only part of the pattern where we use a numbered list.
Is this on purpose? i.e. are these points meant to be used in this order?

fioddor replied:
Arguably yes. I thought of a sequence, but it isn't the only possible one, so I don't have a strong opinion.

spier replied:
Got it. Let's keep it as is then, and see if it sticks :)

spier commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
+  1. document usage of modules.
+  1. introduce semantic versioning
+  1. use a dependency management software such as pip, nuget, npm
+* More information and techniques are available from the book **Working Effectively with Legacy Code** by Michael Feathers and **Refactoring: Improving the Design of Existing Code** by Martin Fowler
```
```suggestion
* More information and techniques are available from the book **Working Effectively with Legacy Code** by Michael Feathers and **Refactoring: Improving the Design of Existing Code** by Martin Fowler.
```

We could add links to the books as well?

- https://martinfowler.com/books/refactoring.html
- https://www.oreilly.com/library/view/working-effectively-with/0131177052/

Never quite sure if links are a good or a bad thing in such cases.

At least the 1st link contains some extra info from the author, i.e. it doesn't "just" sell the book :)

spier replied:
@fioddor wasn't sure if you caught this comment, so reopening it.

Do you think links could be helpful here?

fioddor replied:
[Yes](https://github.com/InnerSourceCommons/InnerSourcePatterns/pull/475/commits/dd08553415c70ff11cb7a24570cfa09703802d6b).

spier reviewed (commented):
@fioddor had time now to step through the changes, so I do understand how the new Solution section came together. Thanks for your pointers to that!

I left a couple more comments for you to consider.

After that this PR is good to be merged from my end.

Commit:
Rewrite success projects

Suggested by reviewer

Co-authored-by: Sebastian Spier <github@spier.hu>

Commit:
Improve style

As suggested by reviewer.

Co-authored-by: Sebastian Spier <github@spier.hu>

fioddor commented on patterns/1-initial/modular-code.md:
```diff
@@ -8,54 +8,71 @@ The lack of modularization in the software architecture prevents reuseability, a
-* There is no corporate mandate for modularized code.
-* This is a new product/new code, not a legacy product/code.
+* There might be potentially reusable functionality buried in monolithic code.
+  Legacy code tends to be less modular than new one.
```
```suggestion
* There might be potentially reusable functionality buried in monolithic code. Legacy code tends to be less modular than new one.
```

spier replied:
✅

Commit:
Conservative markdown.

Reviewer commented:
> Is this meant to be it's own bullet, or should it be part of the previous one?
> 
> If the latter, then we need to check after publishing this whether gitbook renders this correctly. I vaguely recall some behavior there where GitHub and gitbook rendering of markdown differ.

Commit:
Add links to books

As suggested by reviewer.

spier reviewed (approved):
Let's get these changes published!

Great work @fioddor.

Also thanks to you @mishari for doing reviews on earlier versions of this PR.
(if you would like to see your name in the Authors or Acknowledgements section, feel free to add yourself ... via a new PR :))
